content/engine/swarm/networking.md

```diff
@@ -432,6 +444,54 @@ $ docker swarm join \
   192.168.99.100:2377
 ```
 
+## Publish ports on an overlay network
+
+Swarm services connected to the same overlay network effectively expose all
+ports to each other. For a port to be accessible outside of the service, that
+port must be _published_ using the `-p` or `--publish` flag on `docker service
+create` or `docker service update`. Both the legacy colon-separated syntax and
+the newer comma-separated value syntax are supported. The longer syntax is
+preferred because it is somewhat self-documenting.
+
+<table>
+<thead>
+<tr>
+<th>Flag value</th>
+<th>Description</th>
+</tr>
+</thead>
+<tr>
+<td><tt>-p 8080:80</tt> or<br /><tt>-p published=8080,target=80</tt></td>
+<td>Map TCP port 80 on the service to port 8080 on the routing mesh.</td>
+</tr>
+<tr>
+<td><tt>-p 8080:80/udp</tt> or<br /><tt>-p published=8080,target=80,protocol=udp</tt></td>
+<td>Map UDP port 80 on the service to port 8080 on the routing mesh.</td>
+</tr>
+<tr>
+<td><tt>-p 8080:80/tcp -p 8080:80/udp</tt> or <br /><tt>-p published=8080,target=80,protocol=tcp -p published=8080,target=80,protocol=udp</tt></td>
+<td>Map TCP port 80 on the service to TCP port 8080 on the routing mesh, and map UDP port 80 on the service to UDP port 8080 on the routing mesh.</td>
+</tr>
+</table>
+
+## Bypass the routing mesh for a swarm service
+
+By default, swarm services which publish ports do so using the routing mesh.
+When you connect to a published port on any swarm node (whether it is running a
+given service or not), you are redirected to a worker which is running that
+service, transparently. Effectively, Docker acts as a load balancer for your
+swarm services. Services using the routing mesh are running in virtual IP (VIP)
+mode. Even a service running on each node (by means of the `--mode global`
+flag) uses the routing mesh. When using the routing mesh, there is no guarantee
+about which Docker node services client requests.
+
+To bypass the routing mesh, you can start a service using DNS Round Robin
+(DNSRR) mode, by setting the `--endpoint-mode` flag to `dnsrr`. You must run
+your own load balancer in front of the service. A DNS query for the service name
+on the Docker host returns a list of IP addresses for the nodes running the
+service. Configure your load balancer to consume this list and balance the
+traffic across the nodes.
```

Comment:
This section is kinda duplicated with swarm/ingress.md.

Reply:
hmm, almost seems like the two sections are conflicting:

- One says: "To bypass the routing mesh, you must use the long --publish service and set mode to host"
- And the other: "Even a service running on each node (by means of the --mode global flag) uses the routing mesh."

Reply:
I think it's "Even a service running on each node (by means of the --mode global flag) uses the routing mesh"... unless the routing mesh is bypassed?

But the two sections appear to describe different ways to do the bypass, here's it's "start a service using DNS Round Robin (DNSRR) mode". In swarm/ingress it's "use the long --publish service and set mode to host".

(I think they may be related ... if you want to do the load-balancing yourself, set up the service with "host mode", then your load balancer can get the list of servers running it from DNS? But, not sure, @corhere will correct me.)

Reply:
The `--endpoint-mode` option applies to the service as a whole, whereas the publish mode is a per-port option. Each option has two values so there are four possible combinations, [although only three of those are legal.](
https://github.com/moby/swarmkit/blob/5e5ffa8a0b32c027436d5c2fb958a33d1a8bc8d2/manager/controlapi/service.go#L329-L338)

| `--endpoint-mode` | `--publish mode=` | Port published on nodes | Routing |
|--------|--------|--------|--------|
| `dnsrr` | `ingress` | NOT ALLOWED | NOT ALLOWED |
| `dnsrr` | `host` | Nodes where service is running | Same node | 
| `vip` | `ingress` | All nodes in Swarm | Routing mesh |
| `vip` | `host` | ~Nodes where service is running?~ | ~Routing mesh?~ |

As far as I can tell, all that `--endpoint-mode dnsrr` does is disable allocating a Virtual IP on the ingress network for the service and disallow `--publish mode=ingress`. But of course, libnetwork behaves slightly differently depending on whether or not a Virtual IP is allocated. I've used up half a day trying to trace through the code to come up with a straight answer, and I'm very lost. Someone will have to experiment with a running cluster to figure out how vip+host published ports behave.

This section of documentation is trying to convey that when a service using the default `--endpoint-mode vip` has a port published using the default `--publish mode=ingress`, client requests _always_ go through the routing mesh, with no exceptions. It does not matter which nodes the service is running on, or what the service's service mode is set to. The only way to bypass the routing mesh is to configure the service('s published ports?) to bypass the routing mesh.

Reply:
I tried to consolidate the routing mesh bypass to the `ingress.md` page in 67ccdb2298eafee4af02ce3943014a0a4d3d7643 - hopefully I didn't mess it up too bad